File: fm-dddrive/src/main/java/io/dddrive/core/doc/model/base/DocBase.java
package io.dddrive.core.doc.model.base;

import static io.dddrive.util.Invariant.requireThis;

import java.time.OffsetDateTime;
import java.util.List;

import io.dddrive.core.ddd.model.Part;
import io.dddrive.core.ddd.model.base.AggregateBase;
import io.dddrive.core.doc.model.Doc;
import io.dddrive.core.doc.model.DocMeta;
import io.dddrive.core.doc.model.DocPartTransition;
import io.dddrive.core.doc.model.DocRepository;
import io.dddrive.core.doc.model.enums.CodeCaseDef;
import io.dddrive.core.doc.model.enums.CodeCaseStage;
import io.dddrive.core.oe.model.ObjUser;
import io.dddrive.core.property.model.BaseProperty;
import io.dddrive.core.property.model.EnumProperty;
import io.dddrive.core.property.model.PartListProperty;
import io.dddrive.core.property.model.Property;
import io.dddrive.core.property.model.ReferenceProperty;

public abstract class DocBase extends AggregateBase implements Doc, DocMeta {

	//@formatter:off
	protected final BaseProperty<String> docTypeId = this.addBaseProperty("docTypeId", String.class);
	protected final EnumProperty<CodeCaseDef> caseDef = this.addEnumProperty("caseDef", CodeCaseDef.class);
	protected final EnumProperty<CodeCaseStage> caseStage = this.addEnumProperty("caseStage", CodeCaseStage.class);
	protected final BaseProperty<Boolean> isInWork = this.addBaseProperty("isInWork", Boolean.class);
	protected final ReferenceProperty<ObjUser> assignee = this.addReferenceProperty("assignee", ObjUser.class);
	private final PartListProperty<DocPartTransition> transitionList = this.addPartListProperty("transitionList", DocPartTransition.class);
	//@formatter:on

	private CodeCaseStage oldCaseStage;

	protected DocBase(DocRepository<? extends Doc> repository) {
		super(repository);
	}

	@Override
	public DocRepository<?> getRepository() {
		return (DocRepository<?>) super.getRepository();
	}

	@Override
	public DocMeta getMeta() {
		return this;
	}

	@Override
	public Object getTenantId() {
		return this.tenant.getId();
	}

	@Override
	public void doCreate(Object id, Object tenantId, Object userId, OffsetDateTime timestamp) {
		try {
			this.disableCalc();
			super.doCreate(id, tenantId, userId, timestamp);
			this.docTypeId.setValue(this.getRepository().getAggregateType().getId());
		} finally {
			this.enableCalc();
		}
	}

	@Override
	public void doAfterCreate(Object userId, OffsetDateTime timestamp) {
		super.doAfterCreate(userId, timestamp);
		try {
			this.disableCalc();
			this.owner.setId(userId);
			this.version.setValue(0);
			this.createdByUser.setId(userId);
			this.createdAt.setValue(timestamp);
		} finally {
			this.enableCalc();
		}
		// freeze until caseDef is set
		this.freeze(); // TODO reconsider
	}

	@Override
	public void doAfterLoad() {
		super.doAfterLoad();
		this.oldCaseStage = this.getCaseStage();
	}

//	@Override
//	public void doAssignParts() {
//		super.doAssignParts();
//		DocPartItemRepository itemRepository = this.getRepository().getItemRepository();
//		for (Property<?> property : this.getProperties()) {
//			if (property instanceof EnumSetProperty<?> enumSet) {
//				List<DocPartItem> partList = itemRepository.getParts(this, enumSet.getPartListType());
//				enumSet.loadEnums(partList);
//			} else if (property instanceof ReferenceSetProperty<?> referenceSet) {
//				List<DocPartItem> partList = itemRepository.getParts(this, referenceSet.getPartListType());
//				referenceSet.loadReferences(partList);
//			}
//		}
//	}

	@Override
	public void doBeforeStore(Object userId, OffsetDateTime timestamp) {

		DocPartTransitionBase transition = (DocPartTransitionBase) this.transitionList.addPart(null);
		//transition.setSeqNr(this.transitionList.getPartCount() - 1);
		transition.user.setId(userId);
		transition.timestamp.setValue(timestamp);
		transition.oldCaseStage.setValue(this.oldCaseStage);
		transition.newCaseStage.setValue(this.getCaseStage());

		super.doBeforeStore(userId, timestamp);

		try {
			this.disableCalc();
			this.version.setValue(this.version.getValue() + 1);
			this.modifiedByUser.setId(userId);
			this.modifiedAt.setValue(timestamp);
		} finally {
			this.enableCalc();
		}

	}

	@Override
	public boolean isInWork() {
		return this.isInWork.getValue();
	}

	@Override
	public void setCaseDef(CodeCaseDef caseDef) {
		requireThis(this.getMeta().getCaseDef() == null, "caseDef empty");
		requireThis(caseDef != null, "caseDef not null");
		this.unfreeze();
		this.caseDef.setValue(caseDef);
	}

	@Override
	public void setCaseStage(CodeCaseStage caseStage, Object userId, OffsetDateTime timestamp) {
		requireThis(caseStage != null && !caseStage.getIsAbstract(), "valid caseStage (i)");
		requireThis(this.getCaseDef() == null || caseStage.getCaseDef() == this.getCaseDef(), "valid caseStage (ii)");
		if (this.getCaseDef() == null) {
			this.setCaseDef(caseStage.getCaseDef());
		}
		if (this.getCaseStage() == null) { // initial transition
			DocPartTransitionBase transition = (DocPartTransitionBase) this.transitionList.addPart(null);
			//transition.setSeqNr(0);
			transition.user.setId(userId);
			transition.timestamp.setValue(timestamp);
			transition.newCaseStage.setValue(caseStage);
			this.oldCaseStage = caseStage;
		}
		this.caseStage.setValue(caseStage);
		this.isInWork.setValue(caseStage.isInWork());
	}

	@Override
	public List<CodeCaseStage> getCaseStages() {
		return this.getCaseDef().getCaseStages();
	}

	@Override
	public Part<?> doAddPart(Property<?> property, Integer partId) {
		if (property == this.transitionList) {
			return this.getDirectory().getPartRepository(DocPartTransition.class).create(this, property, partId);
		}
		return super.doAddPart(property, partId);
	}

	protected void setCaption(String caption) {
		this.caption.setValue(caption);
	}

//	@Override
//	public void doCalcSearch() {
//		super.doCalcSearch();
//		//Integer orderNr = ((AggregateRepositorySPI<?>) this.getRepository()).getIdProvider().getOrderNr(this.getId());
//		//this.addSearchToken(orderNr + "");
//	}

}

File: fm-dddrive/src/main/java/io/dddrive/core/doc/model/base/DocBase.kt
package io.dddrive.core.doc.model.base

import io.dddrive.core.ddd.model.Part
import io.dddrive.core.ddd.model.base.AggregateBase
import io.dddrive.core.doc.model.Doc
import io.dddrive.core.doc.model.DocMeta
import io.dddrive.core.doc.model.DocPartTransition
import io.dddrive.core.doc.model.DocRepository
import io.dddrive.core.doc.model.enums.CodeCaseDef
import io.dddrive.core.doc.model.enums.CodeCaseStage
import io.dddrive.core.oe.model.ObjUser
import io.dddrive.core.property.model.BaseProperty
import io.dddrive.core.property.model.EnumProperty
import io.dddrive.core.property.model.Property
import io.dddrive.core.property.model.ReferenceProperty
import java.time.OffsetDateTime

abstract class DocBase protected constructor(
	repository: DocRepository<out Doc>,
) : AggregateBase(repository),
	Doc,
	DocMeta {

	// @formatter:off
	protected val _docTypeId: BaseProperty<String> = this.addBaseProperty<String>("docTypeId", String::class.java)
	protected val _caseDef: EnumProperty<CodeCaseDef> = this.addEnumProperty<CodeCaseDef>("caseDef", CodeCaseDef::class.java)
	protected val _caseStage: EnumProperty<CodeCaseStage> = this.addEnumProperty<CodeCaseStage>("caseStage", CodeCaseStage::class.java)
	protected val _isInWork: BaseProperty<Boolean> = this.addBaseProperty<Boolean>("isInWork", Boolean::class.java)
	protected val _assignee: ReferenceProperty<ObjUser> = this.addReferenceProperty<ObjUser>("assignee", ObjUser::class.java)
	private val _transitionList = this.addPartListProperty<DocPartTransition>("transitionList", DocPartTransition::class.java)
	// @formatter:on

	private var oldCaseStage: CodeCaseStage? = null

	override val repository: DocRepository<*>
		get() = super.repository as DocRepository<*>

	override val meta: DocMeta
		get() = this

	override fun doCreate(
		aggregateId: Any,
		tenantId: Any,
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		try {
			this.disableCalc()
			super.doCreate(aggregateId, tenantId, userId, timestamp)
			this._docTypeId.value = this.repository.aggregateType.id
		} finally {
			this.enableCalc()
		}
	}

	override fun doAfterCreate(
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		super.doAfterCreate(userId, timestamp)
		try {
			this.disableCalc()
			this._owner.id = userId
			this._version.value = 0
			this._createdByUser.id = userId
			this._createdAt.value = timestamp
		} finally {
			this.enableCalc()
		}
		// freeze until caseDef is set
		this.freeze() // TODO reconsider
	}

	override fun doAfterLoad() {
		super.doAfterLoad()
		this.oldCaseStage = this.caseStage
	}

	// 	@Override
	// 	public void doAssignParts() {
	// 		super.doAssignParts();
	// 		DocPartItemRepository itemRepository = this.getRepository().getItemRepository();
	// 		for (Property<?> property : this.getProperties()) {
	// 			if (property instanceof EnumSetProperty<?> enumSet) {
	// 				List<DocPartItem> partList = itemRepository.getParts(this, enumSet.getPartListType());
	// 				enumSet.loadEnums(partList);
	// 			} else if (property instanceof ReferenceSetProperty<?> referenceSet) {
	// 				List<DocPartItem> partList = itemRepository.getParts(this, referenceSet.getPartListType());
	// 				referenceSet.loadReferences(partList);
	// 			}
	// 		}
	// 	}

	override fun doBeforeStore(
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		this._transitionList.addPart(null).init(userId, timestamp, oldCaseStage, caseStage!!)
		super.doBeforeStore(userId, timestamp)
		try {
			this.disableCalc()
			this._version.value = this._version.value!! + 1
			this._modifiedByUser.id = userId
			this._modifiedAt.value = timestamp
		} finally {
			this.enableCalc()
		}
	}

	override val isInWork: Boolean get() = this._isInWork.value!!

	override fun setCaseDef(caseDef: CodeCaseDef) {
		require(this.meta.caseDef == null) { "caseDef empty" }
		this.unfreeze()
		this._caseDef.value = caseDef
	}

	override fun setCaseStage(
		caseStage: CodeCaseStage,
		userId: Any,
		timestamp: OffsetDateTime,
	) {
		require(!caseStage.isAbstract) { "valid caseStage (i)" }
		require(this.caseDef == null || caseStage.caseDef === this.caseDef) { "valid caseStage (ii)" }
		if (this.caseDef == null) {
			this.setCaseDef(caseStage.caseDef)
		}
		if (this.caseStage == null) { // initial transition
			this._transitionList.addPart(null).init(userId, timestamp, this.caseStage, caseStage)
			this.oldCaseStage = caseStage
		}
		this._caseStage.value = caseStage
		this._isInWork.value = caseStage.isInWork
	}

	override val caseStages: List<CodeCaseStage>
		get() = this.caseDef?.getCaseStages() ?: emptyList()

	override fun doAddPart(
		property: Property<*>,
		partId: Int?,
	): Part<*> {
		if (property === this._transitionList) {
			return this.directory
				.getPartRepository(DocPartTransition::class.java)
				.create(this, property, partId)
		}
		return super.doAddPart(property, partId)
	}

	protected fun setCaption(caption: String?) {
		this._caption.value = caption
	}

	// 	@Override
	// 	public void doCalcSearch() {
	// 		super.doCalcSearch();
	// 		//Integer orderNr = ((AggregateRepositorySPI<?>) this.getRepository()).getIdProvider().getOrderNr(this.getId());
	// 		//this.addSearchToken(orderNr + "");
	// 	}

}
